[WebGPU] createRender/ComputePipelineAsync are unimplemented

mwyrzykowski opened the pull request:
#### b794bac15380eef65e9a4535759ecf8e9c208b70
<pre>
[WebGPU] createRender/ComputePipelineAsync are unimplemented
<a href="https://bugs.webkit.org/show_bug.cgi?id=253404">https://bugs.webkit.org/show_bug.cgi?id=253404</a>
&lt;radar://106244188&gt;

Reviewed by Myles C. Maxfield.

Add GPUPipelineError and use it to implement createRenderPipelineAsync
and createComputePipelineAsync.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
Add files.

* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::createComputePipelineAsync):
(WebCore::GPUDevice::createRenderPipelineAsync):
Reject the promise on failure as specified in the specification.

* Source/WebCore/Modules/WebGPU/GPUPipelineError.cpp: Added.
(WebCore::GPUPipelineError::GPUPipelineError):
* Source/WebCore/Modules/WebGPU/GPUPipelineError.h: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineError.idl: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineErrorInit.h: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineErrorInit.idl: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineErrorReason.h: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineErrorReason.idl: Added.
New files.

* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
Unified build fix.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createComputePipelineAsync):
(PAL::WebGPU::DeviceImpl::createRenderPipelineAsync):
Return nullptr on failure.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
New files.

* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
New IDLInterface type.

* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipelineAsync):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipelineAsync):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::createComputePipelineAsync):
(WebKit::RemoteDevice::createRenderPipelineAsync):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::createComputePipelineAsync):
(WebKit::WebGPU::RemoteDeviceProxy::createRenderPipelineAsync):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
Return false or nullptr on failure from the callback.

* Source/WebCore/Modules/streams/WritableStream.cpp:
(WebCore::WritableStream::lock):
(WebCore::WritableStream::locked const):
(WebCore::WritableStream::internalWritableStream):
* Source/WebCore/Modules/streams/WritableStream.h:
(WebCore::WritableStream::lock): Deleted.
(WebCore::WritableStream::locked const): Deleted.
(WebCore::WritableStream::internalWritableStream): Deleted.
Unified build fixes.

Canonical link: <a href="https://commits.webkit.org/261415@main">https://commits.webkit.org/261415@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba9f13bf90133e2dab9f66c10f5a869890d05e64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111499 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20635 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3284 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120279 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115558 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22004 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11734 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/2800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16335 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99488 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104299 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98291 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/71 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45101 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13131 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/69 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86905 "13 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13636 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9513 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19081 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52052 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7945 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15603 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
```diff
@@ -29,9 +29,11 @@
 
 ALLOW_UNUSED_PARAMETERS_BEGIN
 ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+ALLOW_COMMA_BEGIN
```
unified build fixes

litherum replied:
Really? Usually unified build fixes are just `#include`s....

mwyrzykowski replied:
```
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/absl/memory/memory.h:33:
/Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/absl/base/macros.h:134:37: note: expanded from macro 'ABSL_HARDENING_ASSERT'
#define ABSL_HARDENING_ASSERT(expr) ABSL_ASSERT(expr)
/Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/absl/base/macros.h:98:3: note: expanded from macro 'ABSL_ASSERT'
  (ABSL_PREDICT_TRUE((expr)) ? static_cast<void>(0) \
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource56.cpp:1:
In file included from /Volumes/Data/OpenSource/Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:33:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/webrtc/api/frame_transformer_interface.h:18:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/webrtc/api/video/encoded_frame.h:18:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/webrtc/api/units/timestamp.h:21:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/webrtc/api/units/time_delta.h:22:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/webrtc/rtc_base/units/unit_base.h:20:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/webrtc/rtc_base/checks.h:58:
/Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/absl/strings/string_view.h:320:43: error: possible misuse of comma operator here [-Werror,-Wcomma]
    return ABSL_HARDENING_ASSERT(!empty()), ptr_[size() - 1];
/Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/absl/strings/string_view.h:320:12: note: cast expression to void to silence warning
    return ABSL_HARDENING_ASSERT(!empty()), ptr_[size() - 1];
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource56.cpp:1:
In file included from /Volumes/Data/OpenSource/Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:33:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/webrtc/api/frame_transformer_interface.h:18:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/webrtc/api/video/encoded_frame.h:17:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/absl/types/optional.h:67:
In file included from /Volumes/Data/OpenSource/WebKitBuild/Debug/usr/local/include/absl/types/internal/optional.h:24:
```

mwyrzykowski replied:
It depends on who includes `Debug/usr/local/include/absl/strings/string_view.h` first with the warning ignored

litherum commented on Source/WebGPU/WebGPU/ComputePipeline.mm:
```diff
@@ -188,10 +188,9 @@ static MTLSize metalSize(auto workgroupSize)
-    UNUSED_PARAM(descriptor);
-    instance().scheduleWork([strongThis = Ref { *this }, callback = WTFMove(callback)]() mutable {
-        callback(WGPUCreatePipelineAsyncStatus_InternalError, ComputePipeline::createInvalid(strongThis), { });
+    auto pipeline = createComputePipeline(descriptor);
```
I think this is fine for now, but we should at least file a bug and add a FIXME with a link to the bug about doing this _actually_ asynchronously

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -474,10 +474,9 @@ static void populateStencilOperation(MTLStencilDescriptor *mtlStencil, const WGP
-    UNUSED_PARAM(descriptor);
-    instance().scheduleWork([strongThis = Ref { *this }, callback = WTFMove(callback)]() mutable {
-        callback(WGPUCreatePipelineAsyncStatus_InternalError, RenderPipeline::createInvalid(strongThis), { });
+    auto pipeline = createRenderPipeline(descriptor);
```
Ditto.

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -226,35 +226,41 @@ void RemoteDevice::createRenderPipeline(const WebGPU::RenderPipelineDescriptor&
-        objectHeap->addObject(identifier, remoteComputePipeline);
-        callback();
+    m_backing->createComputePipelineAsync(*convertedDescriptor, [callback = WTFMove(callback), objectHeap = Ref { m_objectHeap }, streamConnection = m_streamConnection.copyRef(), identifier] (RefPtr<PAL::WebGPU::ComputePipeline>&& computePipeline) mutable {
+        bool result = computePipeline.get();
```
This is so we know whether or not to reject the promise, I assume?

mwyrzykowski replied:
that's correct

litherum reviewed (approved):
Thanks for keeping the patch fairly mechanical 👍

mwyrzykowski commented on Source/WebCore/Modules/streams/WritableStream.cpp:
```diff
@@ -45,6 +46,23 @@ ExceptionOr<Ref<WritableStream>> WritableStream::create(JSC::JSGlobalObject& glo
     return create(globalObject, underlyingSinkValue, strategyValue);
 }
 
+WritableStream::~WritableStream() = default;
```
cc @youennf to verify these unified build fixes are acceptable. They are appearing in https://github.com/WebKit/WebKit/pull/11154 too, I think these changes will fix them there as well

youennf replied:
This seems fine, JSDOMGlobalObject.h is now included in WritableStream.h so no need to include it in WritableStream.cpp

mwyrzykowski committed:
[WebGPU] createRender/ComputePipelineAsync are unimplemented
https://bugs.webkit.org/show_bug.cgi?id=253404
<radar://106244188>

Reviewed by Myles C. Maxfield.

Add GPUPipelineError and use it to implement createRenderPipelineAsync
and createComputePipelineAsync.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
Add files.

* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::createComputePipelineAsync):
(WebCore::GPUDevice::createRenderPipelineAsync):
Reject the promise on failure as specified in the specification.

* Source/WebCore/Modules/WebGPU/GPUPipelineError.cpp: Added.
(WebCore::GPUPipelineError::GPUPipelineError):
* Source/WebCore/Modules/WebGPU/GPUPipelineError.h: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineError.idl: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineErrorInit.h: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineErrorInit.idl: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineErrorReason.h: Added.
* Source/WebCore/Modules/WebGPU/GPUPipelineErrorReason.idl: Added.
New files.

* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
Unified build fix.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createComputePipelineAsync):
(PAL::WebGPU::DeviceImpl::createRenderPipelineAsync):
Return nullptr on failure.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
New files.

* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
New IDLInterface type.

* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipelineAsync):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipelineAsync):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::createComputePipelineAsync):
(WebKit::RemoteDevice::createRenderPipelineAsync):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::createComputePipelineAsync):
(WebKit::WebGPU::RemoteDeviceProxy::createRenderPipelineAsync):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
Return false or nullptr on failure from the callback.

* Source/WebCore/Modules/streams/WritableStream.cpp:
(WebCore::WritableStream::lock):
(WebCore::WritableStream::locked const):
(WebCore::WritableStream::internalWritableStream):
* Source/WebCore/Modules/streams/WritableStream.h:
(WebCore::WritableStream::lock): Deleted.
(WebCore::WritableStream::locked const): Deleted.
(WebCore::WritableStream::internalWritableStream): Deleted.
Unified build fixes.

Canonical link: https://commits.webkit.org/261415@main
